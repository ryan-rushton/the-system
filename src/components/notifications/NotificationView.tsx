import clsx from 'clsx';
import { motion } from 'framer-motion';
import { CloseButton } from './CloseButton';
import { ErrorIcon, InfoIcon, SuccessIcon, WarningIcon } from './Icons';
import styles from './NotificationView.module.scss';
import { ProgressBar } from './ProgressBar';
import { UniqueNotification } from './notifications';

/**
 * A view for a single  This includes the close button and the progress bar.
 * When the progress bar fills the notification will close.
 */
export function NotificationView({
  notification: { severity, message, duration },
  onClose,
}: {
  /** The notification to show. */
  notification: UniqueNotification;
  /** Function to call when the close button is pressed or the duration runs out. */
  onClose(): void;
}) {
  return (
    <motion.li
      data-testid="notification-view"
      className={clsx(styles.notification, {
        [styles.success]: severity === 'success',
        [styles.info]: severity === 'info',
        [styles.warning]: severity === 'warning',
        [styles.error]: severity === 'error',
      })}
      initial={{ opacity: 0, scale: 0.3, y: 50 }}
      animate={{ opacity: 1, scale: 1, y: 0 }}
      exit={{
        opacity: 0,
        scale: 0.5,
        transition: { duration: 0.3 },
      }}
    >
      <div className={styles.content}>
        {severity === 'error' && <ErrorIcon />}
        {severity === 'warning' && <WarningIcon />}
        {severity === 'info' && <InfoIcon />}
        {severity === 'success' && <SuccessIcon />}
        <div className={styles.message}>{message}</div>
        <CloseButton onClose={onClose} />
      </div>
      {duration !== undefined && <ProgressBar duration={duration} onCompletion={onClose} />}
    </motion.li>
  );
}
