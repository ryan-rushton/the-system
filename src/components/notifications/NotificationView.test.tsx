import { cleanup, render, screen } from '@testing-library/react';
import React from 'react';
import NotificationView from './NotificationView';

describe('NotificationView', () => {
  afterEach(cleanup);

  test('it renders', () => {
    render(<NotificationView notification={{ severity: 'error', message: 'Some error!' }} />);

    expect(screen.getByTestId('notification-view')).toBeInTheDocument();
  });

  test('it matches snapshot', () => {
    render(<NotificationView notification={{ severity: 'error', message: 'Some error!' }} />);

    expect(screen.getByTestId('notification-view')).toMatchSnapshot();
  });

  test('it renders the error icon correctly', () => {
    render(<NotificationView notification={{ severity: 'error', message: 'Some error!' }} />);

    expect(screen.getByTestId('error-icon')).toBeInTheDocument();
  });

  test('it renders the warning icon correctly', () => {
    render(<NotificationView notification={{ severity: 'warning', message: 'Some warning!' }} />);

    expect(screen.getByTestId('warning-icon')).toBeInTheDocument();
  });

  test('it renders the info icon correctly', () => {
    render(<NotificationView notification={{ severity: 'info', message: 'Some info!' }} />);

    expect(screen.getByTestId('info-icon')).toBeInTheDocument();
  });

  test('it renders the success icon correctly', () => {
    render(<NotificationView notification={{ severity: 'success', message: 'Some success!' }} />);

    expect(screen.getByTestId('success-icon')).toBeInTheDocument();
  });
});
