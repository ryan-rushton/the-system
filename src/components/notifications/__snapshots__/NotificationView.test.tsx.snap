// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationView > it matches snapshot 1`] = `
<li
  class="_notification_43395a _error_43395a"
  data-testid="notification-view"
  style="opacity: 0; transform: translateY(50px) scale(0.3);"
>
  <div
    class="_content_43395a"
  >
    <div
      class="_icon_caffd4"
      data-testid="error-icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-exclamation fa-lg "
        data-icon="circle-exclamation"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zm0-384c13.3 0 24 10.7 24 24l0 112c0 13.3-10.7 24-24 24s-24-10.7-24-24l0-112c0-13.3 10.7-24 24-24zM224 352a32 32 0 1 1 64 0 32 32 0 1 1 -64 0z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="_message_43395a"
    >
      Some error!
    </div>
    <button
      class="_close_241e48"
      data-testid="close-button"
    >
      <svg
        height="12"
        viewBox="0 0 12 12"
        width="12"
      >
        <path
          d="M 1 1 L 11 11"
          fill="transparent"
          stroke="#fefefe"
          stroke-linecap="round"
          stroke-width="3"
        />
        <path
          d="M 1 11 L 11 1"
          fill="transparent"
          stroke="#fefefe"
          stroke-linecap="round"
          stroke-width="3"
        />
      </svg>
    </button>
  </div>
</li>
`;
